Fix stepper half half step mode

Fractal147 opened the pull request:
##Description:
Half step mode originally had second and third steps mixed up for output A.

**Related issue (if applicable):** fixes <link to issue> [https://github.com/esphome/issues/issues/1655](https://github.com/esphome/issues/issues/1655)



## Checklist:
  - [Y ] The code change is tested and works locally.
  - [N/A ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ N/A] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Fractal147 committed:
Fixed half half step mode

Half step mode originally had second and third steps mixed up for output A.
https://github.com/esphome/issues/issues/1655

stlemme commented:
Thanks for this fix, @Fractal147 
I just faced the same issue and found your esphome/issues#1655 and PR #1397

May I suggest the following code snippet as a replacement which is much more along the lines of the full step implementation and requires no branching at all:
```
case ULN2003_STEP_MODE_HALF_STEP: {
    uint32_t j = i >> 1;
    uint32_t k = i  & 1;

    res |= 1 << j;
    res |= 1 << ((j + k) & 0x3);
    break;
}
 ```
I'm just writing this comment here to avoid having another PRs related to the same issue.

Fractal147 commented:
Nice work @stlemme, no branching is going to be more consistent for timing, if perhaps less readable. Though I think it's always going to be unreadable bit twiddling.
I think it is possible to tweak it a little further, but at this point it's more of a matter of opinion if it makes any real difference once compiled. 

```
case ULN2003_STEP_MODE_HALF_STEP: {
    res |= 1 << (i >>1);
    res |= (res << ( i & 1)) % 15;
    break;
}
```
I'll modify this PR then to have non-branching code, but not terribly used to git!

Fractal147 committed:
Updated half step mode to have no branching

Code based off comments in PR: https://github.com/esphome/esphome/pull/1397#issuecomment-739413973

Fractal147 committed:
removed variable declarations in the switch/case

Oops. No explicit declarations.
Rearranged to be tidier, same output sequence.

Fractal147 committed:
Fixed typo bracket

Minor typo fixed - logic complete to do branchless half stepping.

Fractal147 committed:
Format the brackets to satisfy clang

Fractal147 commented:
Knew I'd mess it up somehow, and manged to do so thrice... 
However, thank you @stlemme, now got something in-between both our codes that should work without branches, and compiles.
